### DISCUSSION TITLE
[백준 1157]  단어 공부

### DISCUSSION THREAD

**Discussion by ChunEunyu** [ghdcksgml1/Algorithm_Study#245] [1일 1알고리즘]
https://www.acmicpc.net/problem/1157

```python
word = input().lower() # 문자열을 입력받아 모두 소문자로 바꾸기
word_list = list(set(word)) # word 리스트에 중복을 제거한 리스트(set함수 이용)
cnt = [] # 중복된 알파벳의 중복 횟수 

for i in word_list: # 중복이 제거된 알파벳 리스트를 반복
    # count 함수는 해당 알파벳이 몇 개 중복되었는지 돌려준다
    count = word.count(i) # 알파벳의 중복 횟수를 count 변수에 저장
    cnt.append(count) # cnt 리스트에 알파벳 중복횟수 저장

if cnt.count(max(cnt)) >= 2: # 가장 개수가 많은 알파벳이 두 개 이상이면
    print("?")
else: # 아니면
    print(word_list[(cnt.index(max(cnt)))].upper()) # 가장 개수가 많은 알파벳을 대문자로 출력
```

set 함수와 count 함수를 배웠다. 
set 함수는 리스트 속 중복된 요소를 제거하여 새로운 리스트를 반환한다.
예를 들어, 
`li = ['a', 'b', 'c', 'a', 'a']
print(set(li))
`
위의 결과값은
{'a', 'c', 'b'}가 된다.

count 함수는 리스트 속 특정 요소가 몇 개 들어있는지 알려준다.
예를 들어,
`li = ['a', 'b', 'c', 'a', 'a']
a = li.count('a')
print(a)`
count 함수를 사용할 땐 인덱스가 아닌 특정 요소를 입력해야한다.

**Comment by ghdcksgml1** [comment]
<img width="85" alt="스크린샷 2022-06-01 오후 8 04 57" src="https://user-images.githubusercontent.com/79779676/171389842-78d4ab86-cf2c-4282-a792-aaa5a61408cb.png">

이렇게 쓰시면 파이썬 코드 입력하실 수 있습니다~

**Reply by ChunEunyu** [reply]
감사합니다~~ discussion에 처음 올려봤는데 공부한 내용도 정리되고 앞으로 자주 올려야겠어요!!

**Reply by ghdcksgml1** [reply]
개인 블로그에 markup 언어로 작성하시고 복붙해서 하는 방법도 좋은방법입니다. 👍

**Comment by BanApp** [comment]
파이썬 유저분이 한분 늘었다니 신나네요!

**Reply by BanApp** [reply]
그리고 파이썬으로 문제 푸실땐 input()으로 받는거 보단 sys모듈에 있는 readline() 쓰시는게 시간부분에서 훨씬 유리합니다!
특히 입력값을 엄청 자주 받는 문제의 경우 input() 때문에 시간초과가 뜨기도 합니다ㅠㅠ
속도는 더 느리지만 count 함수 안쓰고 딕셔너리로 푸는 방법도 있습니다!
파이썬의 조금 어렵거나 복잡한 문자열 문제는 대부분 딕셔너리로 해결이 가능합니다.

```
python

import sys
from string import ascii_uppercase #대문자 라이브러리

word = list(sys.stdin.readline().rstrip().upper()) #sys 모듈로 readline()함수 호출
dic = {}

for i in ascii_uppercase:
    dic[i] = 0

for i in word:
    dic[i] = dic[i] + 1

ans = [i for i, v in dic.items() if max(dic.values()) == v]

if len(ans) == 1:
    print(*ans)
else:
    print('?')


```

**Reply by ChunEunyu** [reply]
오 지금 확인했네요 꿀팁 감사합니다!!! 너무 반갑습니다~~
